### ISSUE TITLE
Persistent partition between boots

### ISSUE THREAD

**Issue by shasheene** [rescuezilla/rescuezilla#8]
Users have [in the past called for a persistent partition](https://sourceforge.net/p/redobackup/discussion/1169664/thread/0e0beed3/), to maintain files between boots.

This could be used to hold network configuration to make using Rescuezilla more efficient.

This is an interesting idea that is worth pursuing. Please comment/react if you would like this feature. Right now, it's very low priority.

**Comment by shasheene** [comment]
A user named "coth" on the comment section of the [AlternativeTo](https://alternativeto.net/software/redo-backup-and-recovery/#post-59665) website wants persistent partitions. I suggested to keep all discussion in this GitHub issue.

> Then, may be, if possible, some edit mode to boot directly from USB without RAM to preinstall apps.

Just like other Linux live environments, Rescuezilla's root filesystem is a SquashFS compressed archive, which is not trivial to modify if Rescuezilla is booted. As of writing, the easiest way to "bake in" new applications into a Rescuezilla ISO image is by adding the package to the build script and building a completely new custom ISO image. This is not easy enough for many home users with customization requirements but may not have the skills to build an ISO from scratch and be satisfied with its correctness.

This goal is achievable with a significant change to Rescuezilla's standard live environment architecture. Instead of an ISO9660 image containing a SquashFS root filesystem, a Rescuezilla image could be very similar to a regular uncompressed Linux hard drive. That is, using an EXT4 root partition. To create an immutable live environment, the the typical boot would then have the [fsprotect](https://packages.ubuntu.com/bionic/admin/fsprotect) enabled via a kernel parameter in the default boot menu item. This would make the root partition mounted as read-only with a copy-on-write RAM overlay. A second boot menu item without the fsprotect kernel option then boots the system in read/write mode without this overlay.

**Comment by regs01** [comment]
Well, i said if possible. As there  are many reasons for that - ability to save settings, ability to install and update apps right on place. But of course there are far more important things - like stable backup and restore in any conditions at the moment. And optional EFI boot.

Still, can you point to how "bake in" apps?

**Comment by shasheene** [comment]
> Still, can you point to how "bake in" apps?

The [build instructions](https://github.com/rescuezilla/rescuezilla/blob/master/docs/build_instructions/BUILD.ISO.IMAGE.md) provides a basic overview. Adding a new package requires first `git clone`'ing the rescuezilla repository, then `git checkout` the most recent tag (1.0.5 right now), then modifying the package list in the file `chroot.steps.part.1.sh` (taking care to preserve the use of trailing backslashes to escape the newlines). Then adding your changes to a new git commit, and finally using `make` to build a brand new ISO.

The resulting ISO image will be the same as the official release, but with slightly newer packages (due to a more recent `apt--get update`). The will also contain a version string like `1.0.5-1-g7db66e`, indicating that it is based on a single commit beyond the v1.0.5, and the build associated is with the git commit SHA 7db66e. This version string is currently written into the same directory where a backup is made, and will be processed by future versions of Rescuezilla so it knows "this backup was created by a v1.0.5 compatible version of Rescuezilla". Other than that, it should be identical to the official release.

For thoroughness I will mention that a custom Rescuezilla ISO image can be analysed using the script `extract_and_checksum_iso.sh` in the root of the git repository, which helps compare it to the official Rescuezilla v1.0.5 with the help of the `meld` diff'ing tool.

Anyway, the process is relatively involved so I think only IT businesses using Rescuezilla heavily will go to the effort to customize the ISO images. Also to anyone reading this, if something like a new package is useful to the general Rescuezilla userbase then please push your commits to your Rescuezilla fork and then create a Pull Request, so it can be merged in and help everybody.

**Comment by shasheene** [comment]
(Please ignore the above "This was referenced to number 108" link above. I mistyped a comment in that issue)

**Comment by shasheene** [comment]
To re-iterate, the currently officially recommended tool to write Rescuezilla images onto USB sticks is [balenaEtcher](https://www.balena.io/etcher/), which simply directly writes the Rescuezilla ISO image file directly onto a USB stick. Rescuezilla boot media made in this way don't have persistent partitions because the filesystems used internally are read-only and the lack of a partition table means new partitions cannot be added.

In this GitHub task I was suggesting to modify the Rescuezilla build scripts to enable persistence, but a [user named Julia has brought to my attention](https://sourceforge.net/p/rescuezilla/discussion/support/thread/0a12c12d39/?limit=250#0ad5) that [Rufus](https://rufus.ie/) **is able to create persistent USB sticks from arbitrary Debian/Ubuntu ISO images**. [From the Rufus ChangeLog.txt](https://github.com/pbatard/rufus/blob/master/ChangeLog.txt):

> o Version 3.7 (2019.09.09)
>    Finalize persistent partition support for Debian and Ubuntu [EXPERIMENTAL]:
>    - Debian with persistence should work out of the box with any recent "Debian Live" ISO.
>    - Ubuntu with persistence should also work *IF* using a post 2019.08.01 ISO. Note that,
>      because of bug #1489855, trying to use persistence with Ubuntu ISOs that were released
>      before August 2019 will only result in rescue mode during boot - You have been warned!
>    - Other distros may work with persistence as long as they use a Debian-like or Ubuntu-like
>      method, and, in the case of Ubuntu-like, if they use casper with the #1489855 bugfix.
>    - Persistence and ext formatting support should still be considered EXPERIMENTAL at this stage.

**This Rufus feature effectively implements this issue but without modifying the build scripts**: Rufus creates a FAT32 root filesystem on the USB stick and extracts the Rescuezilla files, and a storage partition with an ext3 filesystem, then installs GRUB to make the stick bootable.

However, as of writing, the Rufus persistent partition is not usable by the Rescuezilla v2.0 frontend because ext3 filesystem being mounted as /var/log (task )

**Comment by oam7575** [comment]
Based on the 5+ years of comments on this bug : https://bugs.launchpad.net/ubuntu/+source/casper/+bug/1489855
It would seem that USB based persistence might be easier said than done at this point in time for an Ubuntu based distro.

The TLDR is that every other release the bug is fixed and then broken again; then even when its fixed they break something else causing a similar but different regression.

**Comment by shasheene** [comment]
Yeah that's exactly the Launchpad bug referenced in the Rufus v3.7 changelog quoted above.

It's very disappointing to hear that the Ubuntu developers keep re-introducing the bug, because converting Ubuntu-based ISO images like Rescuezilla to persistence USB sticks will be a massive feature!

**Comment by oam7575** [comment]
What if, as part of the build process the scripts create an empty file using fallocate switch to create an empty file.
This would create a file of *size* that takes up practically no space when empty.
Then format the loop file with mkfs.vfat
Then as part of the boot process the file can be mounted  via start up scripts

Not sure how practical this would be; but if it works the resulting change would allow rufus and other programs to write the image directly to USB with no changes on their side as the USB should be Read / Write as FAT32 or NTFS , and the file system would be seeing a normal file that just happens to get larger from time to time.

**Comment by shasheene** [comment]
I've thought about the idea. It might be possible. From the build script:

> * The ISO9660 specification has had a number of extensions over the years. They are important to better understand xorrisofs command-line arguments:
>    * El Torito is the CD-ROM boot record specification, which was released in 1995 (named after the restaurant it was supposedly defined in)
>    * An ISO 9660 filesystem can contain an MBR (Master Boot Record), ESP (EFI System Partition) and GPT (GUID Partition Table) without hampering the (up-to 63) El Torito boot records. The MBR/GPT is stored in the first 32KiB of the filesystem. For developers familiar to hard drive layouts, it may seem like a strange concept to have the MBR partition table stored within filesystem, rather than before the filesystem definition.

The MBR/GPT partition tables are _within_ the ISO9660 filesystem. Rescuezilla's ESP (EFI System Partition) is a 6MB FAT16 filesystem that resides _within_ the ISO9660 filesystem but gets exposed using the partition table.

Due to the design of ISO9660, it's not possible to arbitrarily modify the files in ISO9660 filesystems in-place: the entire ISO9660 filesystem would need to be unpacked then repacked eg, using xorriso. The question is whether it's possible to place the filesystem image you describe immediately _after_ the ISO9660 filesystem, then using the use the ISO9660's MBR/GPT partition tables to point to this. I'm not sure this is possible, but it could be.

By the way, I may be wrong on this but as far as I understand `fallocate` sparse allocation is a operating-system kernel feature: a shortcut to make large allocations quickly. When the ISO image + file is copied to a different machine (say a file-server) for distribution the allocation would be silently "expanded". Not a show stopper, but something to mention: this would suggest that any ISO image + concatenated multi-gigabyte filesystem image of mostly-zeroes would need to be compressed before distribution. Not a show stopper or anything, but I thought I'd mention it.

As somewhat alluded to earlier this GitHub task, it's definitely possible (and actually pretty easy) to completely remove the ISO9660 and SquashFS from the build and instead construct an image in exactly the way a normal desktop Linux environment is setup: MBR/GPT + read/write partitions for a completely persistent image. The only gotcha is you would generally want to "grow" the filesystems on first boot so that your entire eg, 128GB USB stick is utilized rather than only whatever sizes were selected by the build script.

**Comment by GralfR** [comment]
I wonder where to draw the line. I use a live-system to make sure, each time the system boots it has the same state, no matter what I (or any collegue) did last time. Especially concerning backup/restore it's important to have a stable system!
I use rescuezilla like acronis or other imaging software for Windows-PCs to save and restore a stable state of a desktop/laptop. But how often is that done? At maximum once a year (documents are backuped by copy more often).

I do not really need a persistent partition. You can connect to the internet after boot and download/install what you need. Okay, I agee, it would be nice to have the last config available, but in most cases I just want a working environment. No blinky glitter stuff, just basics is the best.

My suggestion:
1) Add the possibility for anybody to easily (by GUI) install the live-system booted from ISO to the HDD. After installation one might change settings and add any software he/she likes to this new system, which might be a virtual machine.
This installation also needs a tool to create a new bootable ISO from that installation by GUI/one-click. This way anybody, who wants to adjust rescuezilla live can create his own live-ISO, even if he is a windows-guy like me.

2) Add a tool to the live-ISO that stores all settings by one click into a file (human-editable, maybe XML?) where the user gives path to (e.g. a USB-drive). So the user can adjust the live-system after boot (change network settings, display resolution, standby settings etc.) and store that easily to a file.
If the system is booted the next time, the defaults are restored, so anybody gets a clean, working system. But a tool on the desktop offers to open/import previously saved settings from external drive (or even an online-cloud?).

That's how I'd like to see it, but I didn't really miss a persistence yet. There are a few tiny things I'd like to change (e.g. disable standby/energy-savings #173), but I can get along so far.

**Comment by shasheene** [comment]
Yep, I expect default behavior for 95% of Rescuezilla uses will remain non-persistence.

I agree it would be ideal to achieve your first point. It would be easier if Rufus was available on Linux, but the [author has no plans to do so](https://github.com/pbatard/rufus/wiki/FAQ#do-you-plan-to-port-rufus-to-linuxmac-ossome-other-os). It could be achieved using an alternative method though.

For your second point, a user was suggesting I look into [Puppy Linux copy-on-write approach](https://www.youtube.com/watch?v=jebMZpEmQlQ) which would basically achieve what you're saying. It's definitely a reasonable approach that works well for Puppy Linux project, but such an approach is not as well suited to Rescuezilla: at this stage the path to long-term maintainability is to keep things as standard as possible by staying with SquashFS and ISO9660, so I'm strongly leaning towards the Rufus approach for persistence.

**Comment by pfrouleau** [comment]
Ventoy may be another solution to achieve this, see https://www.ventoy.net/en/plugin_persistence.html.

**Comment by shasheene** [comment]
That does look really good. I haven't tried Ventoy yet, but I know it's very popular. I will try it and see how it goes.

The [workflow for Windows users](https://youtu.be/bpisA9yh_Eo?t=287) is more complex that the current balenaEtcher instructions, but it might not be too bad.

**Comment by GralfR** [comment]
I wonder if storing to the fist partition is such a good idea. My concerns:

- A live-system should not change the installed system without user-knowledge
- Having the config file on the first partition requires a working partition and known filesystem. If this is not the case (e.g. while restoring the system) I do not have my configs available.
- Can a partition be backup/restored if rescuezilla has read/write-access to a file on that partition?

The ventoy might be okay, if that file is on a USB-drive, so it can be moved to any system I run rescuezilla on.

**Comment by candrapersada** [comment]
This could be used to hold network configuration?
![image](https://user-images.githubusercontent.com/36833664/142750437-84b00c2e-ab8d-4725-ac72-2e072414432d.png)

**Comment by shasheene** [comment]
In future yes. I will add a "remember network credentials" checkbox to the Rescuezilla application. Hopefully one day even a network profile system.

But initially, the persistent partition will only save system network configuration without any change to the Rescuezilla frontend.

**Comment by ehawman-rosenberg** [comment]
A workaround for the moment would be to use Ventoy's Boot Conf Replace plugin https://www.ventoy.net/en/plugin_bootconf_replace.html

By editing your own grub.cfg, you can set anything GRUB can set. I use it for networking stuff, mainly, although it seems many people want resolution changes as well which you can do. For Clonezilla I use ocs_prerun to mount my SMB share. Is there a Rescuezilla equivalent?

**Comment by influx3k** [comment]
This should be priority #1.  Being able to store notes, credentials, etc. is absolutely key.  Hopefully the devs will move this to the front of the to do list.

**Comment by LordZozzy** [comment]
> This should be priority [#1](https://github.com/rescuezilla/rescuezilla/issues/1). Being able to store notes, credentials, etc. is absolutely key. Hopefully the devs will move this to the front of the to do list.

+1
